File: src/main/java/net/zymen/retryablerequests/CacheResponseRequestFilter.java
package net.zymen.retryablerequests;

import org.slf4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

import static org.slf4j.LoggerFactory.getLogger;

class CacheResponseRequestFilter extends OncePerRequestFilter {
    private static final Logger log = getLogger(CacheResponseRequestFilter.class);

    private RequestToChecksumService requestToChecksumService;

    @Autowired
    public CacheResponseRequestFilter(RequestToChecksumService requestToChecksumService) {
        this.requestToChecksumService = requestToChecksumService;
    }

    @Override
    protected void doFilterInternal(HttpServletRequest request,
                                    HttpServletResponse response,
                                    FilterChain filterChain) throws ServletException, IOException {

        String key = requestToChecksumService.requestToChecksum(request);
        log.info("key - {}", key);

        filterChain.doFilter(request, new ResponseRecorderWrapper(response));
    }
}

File: src/main/kotlin/net/zymen/retryablerequests/CacheResponseRequestFilter.kt
package net.zymen.retryablerequests;

import net.zymen.retryablerequests.annotations.RetryableRequest
import net.zymen.retryablerequests.storage.ResponseStorage
import org.slf4j.LoggerFactory
import org.springframework.web.filter.OncePerRequestFilter
import org.springframework.web.method.HandlerMethod
import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

internal class CacheResponseRequestFilter(
        private val requestToChecksumService: RequestToChecksumService,
        private val responseStorage: ResponseStorage,
        private val requestMappingHandlerMapping: RequestMappingHandlerMapping
) : OncePerRequestFilter() {

    override fun doFilterInternal(request: HttpServletRequest,
                                  response: HttpServletResponse,
                                  filterChain: FilterChain) {

        try {
            val handlerExecutionChain = requestMappingHandlerMapping.getHandler(request)

            if (handlerExecutionChain != null) {
                val handlerMethod = handlerExecutionChain.handler

                if (handlerMethod is HandlerMethod &&
                        handlerMethod.hasMethodAnnotation(RetryableRequest::class.java)) {
                    log.info("Has annotation {}", handlerMethod.hasMethodAnnotation(RetryableRequest::class.java))

                    val key = requestToChecksumService.requestToChecksum(request);
                    log.info("key - {}", key);

                    if (responseStorage.hasKey(key)) {
                        replayRecordedResponse(key, response)
                    } else {
                        recordResponse(response, filterChain, request, key)
                    }
                } else {
                    filterChain.doFilter(request, response)
                }
            } else {
                filterChain.doFilter(request, response)
            }
        } catch (e: Exception) {
            log.error("errr", e);
        }
    }

    private fun recordResponse(response: HttpServletResponse, filterChain: FilterChain, request: HttpServletRequest, key: String) {
        val servletResponse = ResponseRecorderWrapper(response)
        filterChain.doFilter(request, servletResponse)

        responseStorage.set(key, servletResponse.recordedContent)
    }

    private fun replayRecordedResponse(key: String, response: HttpServletResponse) {
        val content = responseStorage.get(key)
        log.info("Found content for key ${key} - ${content}")

        response.outputStream.write(content.toByteArray())
        response.addHeader("Content-Type", "application/json")
    }

    companion object {
        private val log = LoggerFactory.getLogger(CacheResponseRequestFilter::class.java)
    }
}
